Mathlib/Analysis/SpecialFunctions/Arcosh.lean

```diff
@@ -0,0 +1,256 @@
+public import Mathlib.Analysis.SpecialFunctions.Log.Basic
+
+/-!
+# Inverse of the cosh function
+
+In this file we define an inverse of cosh as a function from [0, ∞) to [1, ∞).
+
+## Main definitions
+
+- `Real.arcosh`: An inverse function of `Real.cosh` as a function from [0, ∞) to [1, ∞).
+
+- `Real.cosh'`, `Real.arcosh'`: Versions of `Real.cosh` and `Real.arcosh` restricted
+  to the domains [0, ∞) and [1, ∞), respectively.
+
+- `Real.cosh'Equiv`, `Real.cosh'OrderIso`, `Real.cosh'Homeomorph`: `Real.cosh'` as an `Equiv`,
+  `OrderIso`, and `Homeomorph`, respectively.
+
+## Main Results
+
+- `Real.cosh_arcosh`, `Real.arcosh_cosh`: cosh and arcosh are inverse in the appropriate domains.
+
+## Tags
+
+arcosh, arccosh, argcosh, acosh
+-/
+
+@[expose] public section
+
+
+noncomputable section
+
+open Function Filter Set NNReal
+
+open scoped Topology
+
+namespace Real
+
+variable {x y : ℝ}
+
+/-- `arcosh` is defined using a logarithm, `arcosh x = log (x + √(x ^ 2 - 1))`. -/
+@[pp_nodot]
+def arcosh (x : ℝ) :=
+  log (x + √(x ^ 2 - 1))
+
+theorem exp_arcosh (x : ℝ) (hx : 1 ≤ x) : exp (arcosh x) = x + √(x ^ 2 - 1) := by
+  apply exp_log
+  positivity
+
+@[simp]
+theorem arcosh_zero : arcosh 1 = 0 := by simp [arcosh]
+
+lemma x_add_sqrt_x_sq_sub_one_inv (x : ℝ) (hx : 1 ≤ x) :
+  (x + √(x ^ 2 - 1))⁻¹ = x - √(x ^ 2 - 1) := by
+    apply inv_eq_of_mul_eq_one_right
+    rw [← pow_two_sub_pow_two, sq_sqrt (sub_nonneg_of_le (one_le_pow₀ hx)), sub_sub_cancel]
+
+/-- `arcosh` is the right inverse of `cosh` over [1, ∞). -/
+@[simp]
+theorem cosh_arcosh {x : ℝ} (hx : 1 ≤ x) : cosh (arcosh x) = x := by
+  rw [arcosh, cosh_eq, exp_neg, exp_log (by positivity), x_add_sqrt_x_sq_sub_one_inv x hx]
+  ring
+
+@[simp]
+theorem arcosh_eq_zero_iff {x : ℝ} (hx : 1 ≤ x) : arcosh x = 0 ↔ x = 1 := by
+  constructor
+  case mpr => intro hx; subst hx; exact arcosh_zero
+  case mp =>
+  intro h
+  convert congrArg cosh h
+  · rw [cosh_arcosh hx]
+  · rw [cosh_zero]
+
+@[simp]
+theorem sinh_arcosh {x : ℝ} (hx : 1 ≤ x) : sinh (arcosh x) = √(x ^ 2 - 1) := by
+  rw [arcosh, sinh_eq, exp_neg, exp_log (by positivity), x_add_sqrt_x_sq_sub_one_inv x hx]
+  ring
+
+/-- `arcosh` is the left inverse of `cosh` over [0, ∞). -/
+@[simp]
+theorem arcosh_cosh {x : ℝ} (hx : 0 ≤ x) : arcosh (cosh x) = x := by
+rw [arcosh, ← exp_eq_exp, exp_log (by positivity), ← eq_sub_iff_add_eq', exp_sub_cosh,
+    ← sq_eq_sq₀ (sqrt_nonneg _) (sinh_nonneg_iff.mpr hx), ← sinh_sq, sq_sqrt (pow_two_nonneg _)]
+
+@[simp]
+theorem arcosh_nonneg {x : ℝ} (hx : 1 ≤ x) : 0 ≤ arcosh x := by
+  apply log_nonneg
+  calc
+    1 ≤ x := hx
+    _ ≤ x + √(x ^ 2 - 1) := by
+      apply (le_add_iff_nonneg_right _).mpr
+      apply sqrt_nonneg
+
+@[simp]
+theorem arcosh_pos {x : ℝ} (hx : 1 < x) : 0 < arcosh x := by
+  have hx' := le_of_lt hx
+  apply Std.lt_of_le_of_ne (arcosh_nonneg hx')
+  contrapose! hx
+  exact ge_of_eq ((arcosh_eq_zero_iff hx').mp hx.symm).symm
```

Comment:
Essentially the same proof works here.
```suggestion
  apply log_pos
  calc
    1 < x + 0 := by simpa
    _ ≤ x + √(x ^ 2 - 1) := by gcongr; positivity
```
Just for the pedagogical value, I'll also show your approach, but cleaned up the way we would normally write it. For instance, we wouldn't generally apply `Std.lt_of_le_of_ne`. Note the use of dot notation, which can aid simplicity and readability. The lemma `LT.lt_of_ne'` helpfully reverses the direction so you end up with `arcosh x = 0` instead of `0 = arcosh x` at the relevant point.
```lean
  have hx' := hx.le
  apply (arcosh_nonneg hx.le).lt_of_ne'
  contrapose! hx
  exact arcosh_eq_zero_iff hx' |>.mp hx |>.le
```